### PR TITLE
feat: CloudFormation docs in schema

### PR DESCRIPTION
**Pull request by hoffa**
### Description of changes

> Full diff here: https://github.com/aws/serverless-application-model/pull/2816.diff

Adds CloudFormation docs to schema:

Screen Shot 2023-01-26 at 13 39 02

Currently uses heuristics to guess the [documentation page](https://github.com/awsdocs/aws-cloudformation-user-guide/tree/main/doc_source) matching `definitions` in the CloudFormation JSON schema.

It works reasonably well (e.g. `AWS::EC2::CapacityReservationFleet.TagSpecification` corresponds to [`aws-properties-ec2-capacityreservation-tagspecification.md`](https://github.com/awsdocs/aws-cloudformation-user-guide/blob/main/doc_source/aws-properties-ec2-capacityreservation-tagspecification.md)), but there are exceptions such as `AWS::EC2::Instance.Ebs` which correspond to [`aws-properties-ec2-blockdev-template.md`](https://github.com/awsdocs/aws-cloudformation-user-guide/blob/main/doc_source/aws-properties-ec2-blockdev-template.md). 🤔 

There are probably more robust ways of cross-referencing, without depending on a specific format [as provided by GoFormation](https://github.com/awslabs/goformation/blob/master/schema/cloudformation.schema.json), and without heuristics (e.g. using the [CloudFormation resource specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html), which includes a `Documentation` URL which has the correct slug).

The decompressed sizes are relatively large (7 to 11 MB), but all compress nicely (3 MB in total gzipped).

### Description of how you validated changes

Checked diff and tested sample templates in VS Code.

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: bin/add_docs_cfn_schema.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    )
+    args = parser.parse_args()
+
+    schema = json.loads(args.schema.read_text())
+    docs = json.loads(args.docs.read_text())["properties"]
+
+    # Assumes schema is from GoFormation and has consistent structure
+    # TODO: Use a more generic walk
+    # TODO: Use a more robust cross-referencing method than guessing slugs, doesn't always work
+    #       e.g. AWS::EC2::Instance.Ebs corresponds to aws-properties-ec2-blockdev-template
```
Comment (acristin) [inline bin/add_docs_cfn_schema.py:58]: Do we need to special case known exceptions?

File: Makefile [hunk 2]
```diff
@@ -47,10 +47,22 @@ prepare-companion-stack:
 
 update-schema-data:
 	mkdir -p .tmp
+
+	# Update and parse SAM docs
 	rm -rf .tmp/aws-sam-developer-guide
-	git clone https://github.com/awsdocs/aws-sam-developer-guide.git .tmp/aws-sam-developer-guide
+	git clone --depth 1 https://github.com/awsdocs/aws-sam-developer-guide.git .tmp/aws-sam-developer-guide
```
Comment (GavinZZ) [inline Makefile:53]: Why is the `--depth 1` needed here?

Comment (hoffa) [reply Makefile:53]: Faster pull. No need for whole history.

**Review by GavinZZ — APPROVED** [review]

File: bin/add_docs_cfn_schema.py (hunk 1 above)
Comment (hoffa) [reply bin/add_docs_cfn_schema.py:58]: There are more robust ways of doing the cross-referencing, probably using the [CloudFormation resource specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) which includes a `Documentation` link (that seemingly matches the filename in https://github.com/awsdocs/aws-cloudformation-user-guide/tree/main/doc_source) from each property that we can use.

This PR is mostly "good enough" to get the initial iteration out. There are improvements to the process I want to look at, e.g. whether the GoFormation schema is adequate (or whether we should use https://github.com/aws-cloudformation/cloudformation-template-schema or the source [resource provider schemas](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-type-schemas.html)), so don't want to overfit to the current assumptions just yet.

**Review by acristin — APPROVED** [review]

**Review by ssenchenko — APPROVED** [review]

File: bin/add_docs_cfn_schema.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    for def_name, def_schema in schema["definitions"].items():
+        if not def_name.startswith("AWS::"):
+            log(f"Skipping {def_name}: does not start with AWS::")
+            continue
+        # If e.g. AWS::S3::Bucket, we only look under Properties
+        # TODO: Support resource attributes et al.
```
Comment (aaythapa) [inline bin/add_docs_cfn_schema.py:64]: et al?

Comment (hoffa) [reply bin/add_docs_cfn_schema.py:64]: e.g. `Properties` itself, or `Type` and whatnot

Comment (acristin) [reply bin/add_docs_cfn_schema.py:64]:
```
Et al. comes from the Latin phrase meaning “and others.” It is usually styled with a period, but you will occasionally see et al as well. 
```
https://www.merriam-webster.com/dictionary/et%20al. for more details :)

**Review by aaythapa — APPROVED** [review]

**Pull request merged**